client/webserver/site/src/js/wallets.ts

```diff
@@ -595,65 +684,51 @@ export default class WalletsPage extends BasePage {
     page.sendPW.value = ''
     page.sendErr.textContent = ''
-
-    this.showFiatValue(asset.id, 0, page.sendValue)
-    page.sendAvail.textContent = Doc.formatFullPrecision(wallet.balance.available, asset.unitInfo)
-    page.sendLogo.src = Doc.logoPath(asset.symbol)
-    page.sendName.textContent = asset.name
+    this.showFiatValue(assetID, 0, page.sendValue)
+    page.sendAvail.textContent = Doc.formatFullPrecision(wallet.balance.available, ui)
+    page.sendLogo.src = Doc.logoPath(symbol)
+    page.sendName.textContent = name
     // page.sendFee.textContent = wallet.feerate
     // page.sendUnit.textContent = wallet.units
     box.dataset.assetID = String(assetID)
-    this.animation = this.showBox(box, page.walletPass)
+    this.showForm(box)
   }
 
   /* doConnect connects to a wallet via the connectwallet API route. */
   async doConnect (assetID: number) {
     const loaded = app().loading(this.body)
-    const res = await postJSON('/api/connectwallet', {
-      assetID: assetID
-    })
+    const res = await postJSON('/api/connectwallet', { assetID })
     loaded()
     if (!app().checkResponse(res)) return
-    const rowInfo = this.rowInfos[assetID]
-    Doc.hide(rowInfo.actions.connect)
+    this.updateDisplayedAsset(assetID)
   }
 
-  /* createWalletSuccess is the success callback for wallet creation. */
-  async createWalletSuccess (assetID: number) {
-    const rowInfo = this.rowInfos[assetID]
-    this.showMarkets(rowInfo.assetID)
-    const a = rowInfo.actions
-    Doc.hide(a.create, a.createPending)
-    Doc.show(a.send, a.deposit)
-    if (assetIsConfigurable(assetID)) Doc.show(a.settings)
-    await app().fetchUser()
-    await app().loadPage('wallets')
+  /* openWalletSuccess is the success callback for wallet unlocking. */
+  async openWalletSuccess (assetID: number, form?: PageElement) {
+    this.assetUpdated(assetID, form, intl.prep(intl.ID_WALLET_UNLOCKED))
   }
 
-  /* openWalletSuccess is the success callback for wallet unlocking. */
-  async openWalletSuccess () {
-    const rowInfo = this.rowInfos[this.openAsset]
-    const a = rowInfo.actions
-    Doc.show(a.send, a.deposit)
-    Doc.hide(a.unlock, a.connect)
-    if (app().walletMap[rowInfo.assetID].encrypted) {
-      Doc.show(a.lock)
+  assetUpdated (assetID: number, oldForm?: PageElement, successMsg?: string) {
+    if (assetID !== this.selectedAssetID) return
+    this.updateDisplayedAsset(assetID)
+    if (oldForm && Object.is(this.currentForm, oldForm)) {
+      if (successMsg) this.showSuccess(successMsg)
+      else this.closePopups()
     }
-    this.showMarkets(this.openAsset)
   }
 
   /* send submits the send form to the API. */
-  async send () {
+  async send (): Promise<void> {
     const page = this.page
     Doc.hide(page.sendErr)
-    const assetID = parseInt(page.sendForm.dataset.assetID || '')
-    const subtract = page.subtractCheckBox.checked || false
+    const assetID = parseInt(page.sendForm.dataset.assetID ?? '')
```

Comment:
Eeeek have we broadly misusing `||` when we should have been using the nullish coalescing operator?

Reply:
Meh. I just discovered this, really. Will try to use it more in the future, but it's already bitten me. I did something like

```Javascript
const v = parseFloat(inputElement.value ?? '0')
```

which yields `NaN`, because `value` is not `undefined` or `null`, just empty, which is falsy but not nullish.

I think we should use it, but I'm not going to enforce it **too** strictly.

---

I've also been using the **optional chaining operator**, `?.`.

```Javascript
const thing = { a: 1 }
let v = thing.b.c    // error thrown cuz b undefined
v = thing?.b?.c      // undefined

```